Feature/#6 #7 #8 api bringup

Benjamin8282 opened the pull request:
이 풀 리퀘스트는 날씨 데이터 조회 기능을 프로젝트에 통합하기 위한 주요 기능들을 추가합니다. 이번 변경사항에는 API 응답을 위한 DTO 추가, Retrofit과 OkHttp를 활용한 네트워크 모듈 구성, 날씨 API 인터페이스 구현, 그리고 API 테스트 코드 작성이 포함되어 있습니다. 또한 `local.properties` 파일을 활용하여 날씨 API 키를 안전하게 관리합니다.

### 🌐 API 연동 및 네트워크 구성

* **API 응답 DTO 추가**  
  `WeatherResponse`, `ForecastResponse` 등 OpenWeatherMap API 응답을 매핑하기 위한 데이터 클래스를 `app/src/main/java/com/ben/simpleweather/data/remote/dto/` 디렉토리에 추가했습니다. 공통으로 재사용되는 구성 요소(`MainDto`, `WeatherDto` 등)도 포함되어 있습니다.  
  ([보기](diffhunk://#diff-0e6c4d4e363a98225919e491ade7c2d21970b1926f9040e44530dff7f1814431R1-R47), [보기](diffhunk://#diff-1b7481216325b1d63ba9215d0b1bbe2dca4b5fb51211a5a1bf0d1488f9c17b97R1-R94))

* **네트워크 모듈 구현**  
  `NetworkModule.kt` 파일을 추가하여 Retrofit, Kotlin Serialization, OkHttp 설정을 구성했습니다. 모든 요청에 API 키가 자동으로 포함되도록 인터셉터를 추가했습니다.  
  ([app/src/main/java/com/ben/simpleweather/di/NetworkModule.ktR1-R71](diffhunk://#diff-01a49f19ad97f585db964f7ddbab2be69800e0ff63d864eeaf0163c591c336d0R1-R71))

* **날씨 API 인터페이스 생성**  
  현재 날씨 및 예보 정보를 가져오기 위한 `WeatherApi` 인터페이스를 정의했습니다.  
  ([app/src/main/java/com/ben/simpleweather/network/WeatherApi.ktR1-R26](diffhunk://#diff-c1e6d68d92879b66a6fd4f490871535a37fe2ec0a61090fe96d7a88b3291dad2R1-R26))

### ⚙️ 설정 및 빌드 구성

* **API 키 관리 추가**  
  `local.properties` 파일에서 API 키를 로드하고, `build.gradle.kts`를 수정하여 빌드 설정에 주입되도록 구성했습니다.  
  ([보기](diffhunk://#diff-8cff73265af19c059547b76aca8882cbaa3209291406f52df1dafbbc78e80c46R1-R2), [보기](diffhunk://#diff-8cff73265af19c059547b76aca8882cbaa3209291406f52df1dafbbc78e80c46R28-R31))

### ✅ 테스트

* **API 통합 테스트 추가**  
  실제 API 호출을 통해 `WeatherApi`의 동작을 검증하는 테스트 코드(`WeatherApiTest`)를 작성했습니다. 응답값에 대한 검증도 포함되어 있습니다.  
  ([app/src/test/java/com/ben/simpleweather/network/WeatherApiTest.ktR1-R102](diffhunk://#diff-9ef65b5ace4812f57dd331f160475638fdf450eca0089047956078b904836fa0R1-R102))

---

closes #6  
closes #7  
closes #8

Commit:
feat: 날씨 예보 응답 DTO 추가 (#8)

날씨 예보 API 응답을 처리하기 위한 `ForecastResponse.kt` 파일과 관련 데이터 클래스들을 추가했습니다. 이 DTO들은 예보 항목, 도시 정보 등을 포함하며, 기존 DTO의 일부를 재사용합니다.

Commit:
feat: 날씨 API 응답 DTO 추가 (#7)

날씨 API의 응답을 처리하기 위한 데이터 클래스들을 추가했습니다. `WeatherResponse`를 비롯하여 `CoordDto`, `WeatherDto`, `MainDto`, `WindDto`, `CloudsDto`, `RainDto`, `SnowDto`, `SysDto`가 포함됩니다.

Commit:
feat: 네트워크 모듈 및 WeatherApi 인터페이스 추가 (#6)(#7)(#8)

Hilt를 사용한 네트워크 모듈(NetworkModule.kt)을 추가하여 Retrofit, OkHttpClient, Kotlinx Serialization Json 인스턴스를 제공합니다. 또한, OpenWeatherMap API 통신을 위한 WeatherApi 인터페이스(WeatherApi.kt)를 정의하여 현재 날씨 및 예보 정보를 가져오는 함수를 포함시켰습니다.

Commit:
feat: API 키 자동 추가 및 BuildConfig 설정 (#6)

OkHttpClient 인터셉터를 사용하여 모든 API 요청에 자동으로 API 키(`appid`)를 쿼리 파라미터로 추가하도록 `NetworkModule.kt`를 수정했습니다. 또한, `build.gradle.kts` 파일을 수정하여 `local.properties` 파일에 저장된 `WEATHER_API_KEY`를 BuildConfig를 통해 사용할 수 있도록 설정했습니다. 이에 따라 `WeatherApi.kt` 인터페이스의 API 키 파라미터가 제거되었습니다.

Commit:
test: WeatherApi 테스트 코드 추가 (#6)(#7)(#8)

OpenWeatherMap API의 현재 날씨 및 예보 API 연동을 검증하는 `WeatherApiTest.kt` 파일을 추가했습니다. 이 테스트는 서울의 위도와 경도를 사용하여 실제 API 호출을 수행하고, 응답 데이터의 유효성을 검사합니다.

주요 테스트 항목:
- `getCurrentWeather`: 현재 날씨 정보 (응답 코드 200, 날씨 목록, 주요 데이터, 도시 이름)
- `getWeatherForecast`: 날씨 예보 정보 (응답 코드 "200", 예보 목록, 도시 데이터, 도시 이름)

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/data/remote/dto/ForecastResponse.kt:
```diff
@@ -0,0 +1,47 @@
+    val message: Int? = null,
+    val cnt: Int? = null,
+    val list: List<ForecastItemDto>? = null,
+    val city: CityDto? = null
```
data class 마다 trailing comma 붙여 주세요.

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/data/remote/dto/WeatherResponse.kt:
```diff
@@ -0,0 +1,94 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class WeatherResponse(
```
마찬가지로 data class 마다 trailing comma 붙여 주세요

f-lab-Toru commented on app/src/main/java/com/ben/simpleweather/network/WeatherApi.kt:
```diff
@@ -0,0 +1,26 @@
+
+    @GET("data/2.5/weather")
+    suspend fun getCurrentWeather(
+        @Query("lat") latitude: Double,
```
이렇게 Query 를 명시적으로 하나하나 해도 되고, QueryMap 을 써도 됩니다.

Benjamin8282 replied:
넵 QueryMap 도 공부 해 보겠습니다

f-lab-Toru reviewed (commented):
수고했습니다

Commit:
test: WeatherApiTest 테스트 코드 수정 (#6)(#7)(#8)

`WeatherApiTest.kt`에서 불필요한 주석과 `println` 문을 제거하여 코드를 정리했습니다.

Commit:
refactor: `WeatherApiTest.kt`에서 불필요한 `println` 제거 (#6)(#7)(#8)

`getWeatherForecast` 테스트에서 날씨 예보 응답을 출력하던 `println` 구문을 제거했습니다.

Commit:
fix: `local.properties`에 `WEATHER_API_KEY` 누락 시 빌드 오류 발생(#6)

`app/build.gradle.kts` 파일을 수정하여 `local.properties` 파일에 `WEATHER_API_KEY`가 없거나 비어있는 경우 빌드 시 GradleException을 발생시키도록 변경했습니다. 이를 통해 API 키 누락으로 인한 런타임 오류를 사전에 방지합니다.

Commit:
style: DTO 클래스 trailing comma 추가(#7)(#8)

`WeatherResponse.kt` 및 `ForecastResponse.kt` 파일의 모든 DTO 클래스 생성자 파라미터에 trailing comma를 추가했습니다.

Commit:
fix: `MainDto`의 `temp_kf` 필드에 `@SerialName` 어노테이션 추가 (#7)(#8)

`WeatherResponse.kt` 파일의 `MainDto` 데이터 클래스에서 `temp_kf` 필드에 `@SerialName("temp_kf")` 어노테이션을 추가하여 JSON 직렬화/역직렬화 시 필드 이름을 명시적으로 지정하도록 수정했습니다.

Commit:
test: `WeatherApiTest` 테스트 함수명 변경 및 API 키 검증 로직 개선(#6)(#7)(#8)

`WeatherApiTest.kt` 파일에서 테스트 함수의 이름을 보다 명확하게 변경했습니다. 예를 들어, ``getCurrentWeather for Seoul should return current weather data``는 `getCurrentWeatherForSeoulShouldReturnCurrentWeatherData`로 변경되었습니다. 또한, `setUp` 함수에서 `BuildConfig.WEATHER_API_KEY`가 비어 있는지 확인하는 로직을 `check` 함수를 사용하도록 수정하여 가독성을 향상시켰습니다.
